### PR TITLE
http.sys; add opt-in support for kernel-mode response buffering

### LINKED ISSUE #5852: Kernel response buffering
Http.Sys supports internal response buffering for better response performance.  WebSockets already enables this by default.  However, enabling this prevents the developer from doing multiple concurrent writes.

Consider enabling this by default, and allowing it to be configured per server and/or per-request.

### LINKED ISSUE #14455: Enable kernel response buffering for HTTP.sys server
I deploy an ASP.NET Core Service in Service Fabric in Azure.
I tried Kestrel and Http.sys as server and using Response compression.
I found if I use Http.sys as server and usign Response compression, static file download is very slow It will cost more than 1 minutes to doanload a 1.6mb JS file.
If I use Kestrel, this issue will not happen. If I use http.sys without Response compression, it also did not happen.
So maybe something wrong with Http.sys and Response compression?

I tried it on ASP.Net 2.2 and 3.0, it can repro in both environments. 
### To Reproduce
Building an ASP.NET Core Service with Http.sys as server (using https), 
Using "app.UseResponseCompression();" to compress the static file response. Then deploy it to Azure Service Fabric with 5 nodes.

Using browsers to open website. It will take a long time to download static files.
If using Kestrel or  Http.sys without compression, it will not happen.
And if I run this service on my local machine, it will never happen Whatever combination.

### Screenshots
image

### LINKED ISSUE #47777: API http.sys: EnableKernelResponseBuffering
## Background and Motivation

Additional configuration option for http.sys : https://github.com/dotnet/aspnetcore/pull/47776

## Proposed API

```diff
namespace Microsoft.AspNetCore.Server.HttpSys;

public class HttpSysOptions
{
+    public bool EnableKernelResponseBuffering { get; set; } 
}
```

## Usage Examples

```csharp
builder.UseHttpSys(options =>
{
    options.EnableKernelResponseBuffering = true;
});
```

## Alternative Designs

for back-port and compat; app-context switch that also impacts initial value

## Risks

None; no change if not used; no overload resolution issues.

### PR DESCRIPTION
**Pull request by mgravell**
# http.sys; add opt-in support for kernel-mode response buffering

Adds support in for buffered responses when using http.sys, to avoid performance problems in small-write/high-latency scenarios.

## Description

As per linked issue, when using http.sys, in some scenarios, high volumes of small writes with high latency may cause significant performance impact, due to lack of a `Pipe` buffer in the http.sys implementation; as a simple workaround, here we optionally enable the http.sys support for response buffering as a global flag.

This is exposed by a new `HttpSysOptions.EnableKernelResponseBuffering` option.

This work is based on the very old (and not mergeable) work in [#418](https://github.com/aspnet/HttpSysServer/pull/418), with some additional input - credit to @davidni 

It is also planned to back-port this feature to net6/net7 (with the property `internal` - no change to public API), via an app-context switch `"Microsoft.AspNetCore.Server.HttpSys.EnableKernelResponseBuffering"`; <strike>this switch detection is retained in the default value to avoid issues for consumers when upgrading.</strike>

We have briefly discussed some new feature to allow per-request enable/disable, but have not implemented that at the current time. We can revisit that in the future.

Also considered: implementing a full `Pipe` implementation in the http.sys code with the pipe consumer writing to http.sys; attempting to minimise code impact until we have a definite demand for that.

Fixes #14455

Fixes [#5852](https://github.com/dotnet/aspnetcore/issues/5852)

Impact is hard to benchmark in a local network - latency is key; however, the change has been successfully validated internally via a private build. Ask me for details if you're curious. As a summary of the validation scenario - a large payload streaming operation that reads (proxies) an external request and mirrors that to response writes:

- without this change: takes over 3 minutes and eventually fails (not having transferred all the information)
- with this change: completes successfully in 8-10 seconds

### CODE REVIEW

**Comment by mgravell** [comment]
OK bot; have it your way; API change submitted separately

**Comment by davidni** [comment]
Happy to see this finally making it to supported releases, thanks @mgravell!

File: src/Servers/HttpSys/src/HttpSysOptions.cs [hunk 1]
```diff
@@ -110,6 +114,15 @@ public HttpSysOptions()
     /// </summary>
     public bool ThrowWriteExceptions { get; set; }
 
+    /// <summary>
+    /// Enable buffering of response data in the Kernel.
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/HttpSysOptions.cs:118]: + The default is false.

File: src/Servers/HttpSys/src/HttpSysOptions.cs [hunk 2]
```diff
@@ -34,6 +34,10 @@ public class HttpSysOptions
     /// </summary>
     public HttpSysOptions()
     {
+        // this feature is being back-ported to net6/net7 via an app-context switch; respect that even on net8+ to avoid
+        // surprises when upgrading TFMs
+        const string EnableKernelResponseBufferingSwitch = "Microsoft.AspNetCore.Server.HttpSys.EnableKernelResponseBuffering";
+        EnableKernelResponseBuffering = AppContext.TryGetSwitch(EnableKernelResponseBufferingSwitch, out var enabled) && enabled;
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/HttpSysOptions.cs:40]: It's standard practice to remove app context switches in the next major release (8.0). They're only used for servicing.

**Comment by mgravell** [comment]
@davidni thanks for your original work which - the above is *almost* exactly that (it was completely impossible to merge the original PR, sorry - but I did try to at least acknowledge credit!) - for your interest, only minor tweak was to ensure that *all* writes use the flag - to avoid http.sys getting confused; the main problem I had here was validating it!

**Review by Tratcher — APPROVED** [review]
Can you at least summarize the perf improvements observed?

**Comment by BrennanConroy** [comment]
> Can you at least summarize the perf improvements observed?

And has this impacted other scenarios besides the high latency + small writes + large payload scenario?

**Comment by mgravell** [comment]
> Can you at least summarize the perf improvements observed?

added in main post

**Comment by BrennanConroy** [comment]
From the docs https://learn.microsoft.com/windows/win32/api/http/nf-http-httpsendresponseentitybody
> Applications using asynchronous I/O which may have more than one send outstanding at a time should not use this flag.

Does this mean the entire application can't have concurrent sends? Are we preventing overlap when this is set? Am I misunderstanding the docs?

**Comment by amcasey** [comment]
@mgravell What's left here?

**Comment by JamesNK** [comment]
Blocked on API review: https://github.com/dotnet/aspnetcore/issues/47777

**Comment by mgravell** [comment]
@BrennanConroy we already block (synchronously or asynchronously as appropriate) individual writes - concurrent *parallel* writes and overlapped (i.e. not awaiting a WriteAsync) is always going to be unsupported, regardless of API or server implementation. Which is a long way of saying "no change here".

**Comment by davidni** [comment]
Side-note for posterity -- when I played with HTTP.sys perf on Asp .NET Core back in the day, one of our experiments was to NOT enable HTTP.sys buffering, and instead to do overlapped writes, ensuring there were always at least N outstanding writes. This helped improve perf in our scenarios (proxying large responses from a destination service -- before YARP), but (surprisingly at the time) we were never able to achieve the same throughput that http.sys kernel buffering afforded.

**Review by halter73 — APPROVED** [review]

**Comment by davidni** [comment]
This is no longer blocked, is it? Per [#47777](https://github.com/dotnet/aspnetcore/issues/47777#issuecomment-1544304751), the proposed api change is reviewed and approved.

Sugg removing the `blocked` tag.

**Pull request merged**